doc/dev/kdb-add-new-command.md

```diff
@@ -0,0 +1,170 @@
+	ksAppendKey (spec, keyNew (COMMAND_SPEC_KEY (COMMAND_NAME) "/all", KEY_META, "description", "Consider all of the keys", KEY_META,
+				   "opt", "a", KEY_META, "opt/long", "all", KEY_META, "opt/arg", "none", KEY_END));
+	ksAppendKey (spec, keyNew (COMMAND_SPEC_KEY (COMMAND_NAME) "/name", KEY_META, "description", "The name of the key", KEY_META,
+				   "args", "indexed", KEY_META, "args/index", "0", KEY_END));
+
+	ADD_BASIC_OPTIONS (spec, COMMAND_SPEC_KEY (COMMAND_NAME))
+}
+```
+
+The `ADD_BASIC_OPTIONS` at the end adds options that are common across all commands to its spec, this includes things like `verbose: -v` or `debug: -d`.
+It if your command should support these, it makes sense to use the macro.
+
+#### Execution
+
+The `int exec... (KeySet * options, Key * errorKey)` receives two parameters, `options` and `errorKey`.
+Where `options` contains the passed options (as parsed by `elektraGetOpts`) and arguments and `errorKey` is where any errors or warnings have to be written to so the framework can properly print them.
+It should start with the `GET_BASIC_OPTIONS` macro, this sets everything up needed for logging(and logging level), basic options and colored printing.
+So afterward `CLI_PRINT` and `CLI_ERROR_PRINT` can be used.
+When returning for the `exec` function the `RETURN (n)` macro should be used, it makes sure things allocated by `GET_BASIC_OPTIONS` are freed properly.
+A basic version of this function could look something like this:
+
+```c
+int ret = 0;
+	GET_BASIC_OPTIONS
+
+	bool all = false;
+	tmp = GET_OPTION_KEY (options, "all");
+	if (tmp != NULL)
+	{
+		elektraKeyToBoolean (GET_OPTION_KEY (options, "all"), &all);
+		keyDel (tmp);
+	}
+
+    ...
+
+	Key * toLookUp = getKeyFromOptions (GET_OPTION (options, "name"), errorKey, verbose);
+	if (toLookUp == NULL)
+	{
+		RETURN (2)
+	}
+
+    ...
+
+	if (found == NULL)
+	{
+		CLI_ERROR_PRINT (CLI_LOG_NONE, "Did not find key '%s'", RED (keyName (toLookUp)));
+		ret = 11;
+		goto cleanup;
+	}
+
+	if (keyGetNamespace (found) == KEY_NS_DEFAULT)
+	{
+		CLI_PRINT (CLI_LOG_VERBOSE, "The key was not found in any other namespace, taking the %s\n", BOLD ("default"));
+	}
+
+	...
+
+cleanup:
+	...
+
+	RETURN (ret)
+```
+
+For getting keys from arguments the helper function `getKeyFromOptions` can the used.
+It will resolve bookmarks and make sure the argument is a valid keyname.
+
+## 2. Output
+
+All output in the `exec` function should be done using either `CLI_PRINT` or `CLI_ERROR_PRINT` with the appropriate logging level.
+Available logging levels are `CLI_LOG_NONE`, `CLI_LOG_VERBOSE` and `CLI_LOG_DEBUG`.
+By using `CLI_PRINT` or `CLI_ERROR_PRINT` logging levels are handled automatically.
+
+## 3. Errors/Warning
+
+If errors occur during the execution of your command, they should be set in `errorKey` and not printed directly.
+This could look something like this:
+
+```c
+ELEKTRA_SET_INTERNAL_ERROR (errorKey, "could not copy key meta to errorKey");
```

Comment:
Doing like this we don't know which errors are from Elektra and which are from CLI? Probably we should provide ELEKTRA_SET_CLI_ERROR?